File: CinemaReservations/CSharp/CinemaAllocations.Infra.RestApi/Controllers/MovieScreeningController.cs
using System;
using CinemaAllocations.Domain;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;

namespace CinemaAllocations.Infra.RestApi.Controllers
{
    [ApiController]
    public class MovieScreeningController : ControllerBase
    {
        private readonly ILogger<MovieScreeningController> _logger;
        private readonly TicketBooth _ticketBooth;

        public MovieScreeningController(ILogger<MovieScreeningController> logger, TicketBooth ticketBooth)
        {
            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
            _ticketBooth = ticketBooth ?? throw new ArgumentNullException(nameof(ticketBooth));
        }

        [HttpPost]
        [Route("moviescreening/{showId}/allocateseats/{partyRequested}")]
        public IActionResult AllocateSeats([FromRoute] string showId, [FromRoute] int partyRequested)
        {
            try
            {
                var allocatedSeats = _ticketBooth.AllocateSeats(
                    new AllocateSeats(showId, partyRequested)
                );

                switch (allocatedSeats)
                {
                    case NoPossibleAllocationsFound noPossibleAllocationsFound:
                        return NotFound(noPossibleAllocationsFound);
                    case TooManyTicketsRequested tooManyTicketsRequested:
                        return BadRequest(tooManyTicketsRequested);
                    case NoPossibleAdjacentSeatsFound noPossibleAdjacentSeatsFound:
                        return NotFound(noPossibleAdjacentSeatsFound);
                    default:
                        return new OkObjectResult(new Dto.SeatsAllocated(allocatedSeats));
                }
            }
            catch (Exception exception)
            {
                _logger.LogError(
                    5001,
                    exception,
                    "Unexpected exception, outside of our domain model. ShowId: {0}, PartyRequested {1}",
                    showId, partyRequested);
                return StatusCode(StatusCodes.Status500InternalServerError,
                    "Unexpected exception. Stay tune, we are sending pigeons!");
            }
        }
    }
}

File: CinemaReservations/Java/CinemaAllocationsInfra/src/main/java/com/sdd/cinemaallocationsinfra/controllers/MovieScreeningController.java
package com.sdd.cinemaallocationsinfra.controllers;

import com.sdd.cinemaallocations.*;
import com.sdd.cinemaallocationsinfra.controllers.model.SeatsAllocated;
import com.sdd.cinemaallocationsinfra.repository.JPAMovieScreeningRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("api/moviescreening")
public class MovieScreeningController {

    private TicketBooth ticketBooth;

    public MovieScreeningController(JPAMovieScreeningRepository jpaMovieScreeningRepository) {
        ticketBooth = new TicketBooth(jpaMovieScreeningRepository);
    }

    @PostMapping(value = "/{showId}/allocateseats/{partyRequested}", produces = "application/json")
    public ResponseEntity<SeatsAllocated> allocateSeats(@PathVariable String showId, @PathVariable Integer partyRequested) {
        AllocateSeats allocateSeats = new AllocateSeats(showId, partyRequested);
        com.sdd.cinemaallocations.SeatsAllocated seatsAllocated = ticketBooth.allocateSeats(allocateSeats);
        if (seatsAllocated instanceof TooManyTicketsRequested) {
            return ResponseEntity.badRequest().build();
        } else if (seatsAllocated instanceof NoPossibleAdjacentSeatsFound) {
            return ResponseEntity.notFound().build();
        } else if (seatsAllocated instanceof NoPossibleAllocationsFound) {
            return ResponseEntity.notFound().build();
        } else {
            return ResponseEntity.ok(SeatsAllocated.fromDomain(seatsAllocated));
        }
    }

}
